CorsixTH/Lua/rooms/training.lua

```diff
@@ -74,21 +68,49 @@ function TrainingRoom:roomFinished()
   Room.roomFinished(self)
 end
 
-function TrainingRoom:calculateTrainingFactor(skeletons, bookcases)
-  -- TODO: tweak/change this function, used in Staff:trainSkill(...)
+--! Determing the training factor of this room for teaching specialisms
+--!param objects (table) the list of objects in this room
+--!return total (number) The final training factor
+function TrainingRoom:calculateTrainingFactor(objects)
+  -- Tally relevant objects that affect training
+  local function countTrainingObjects()
+    local projectors = 0 -- There should only ever be one
+    local skeletons = 0
+    local bookcases = 0
+    for object, _ in pairs(objects) do
+      if object.object_type.id == "skeleton" then
+        skeletons = skeletons + 1
+      elseif object.object_type.id == "bookcase" then
+        bookcases = bookcases + 1
+      elseif object.object_type.id == "projector" then
+        projectors = projectors + 1
+      end
+    end
+    return {[0] = projectors, bookcases, skeletons} -- [0] to align with level_config
```

Comment:
If you make a table with the object names, the loop at line 80 becomes simpler.

Reply:
I understood the idea of matching `[0]` here with the `level_config.gbv.TrainingValue[0]`. However, it doesn't bring any advantages, you still write all 3 lines (if you used a loop for example, you'd gain with it (not sure it's worthwhile though).

The alternative is what I had in mind, except with a bit less duplicated `object.object_type.id` :)  That also helps in reducing needed width.

Untested, but perhaps
``` lua
for object, _ in pairs(objects) do
  local obj_id = object.object_type.id
  counts[obj_id] = counts[obj_id] and counts[obj_id] + 1 or nil
end
```